python/pyspark/sql/connect/client/reattach.py

```diff
@@ -0,0 +1,166 @@
+        first_try = True
+        ret: Optional[pb2.ExecutePlanResponse] = None
+        for attempt in Retrying(can_retry=SparkConnectClient.retry_exception, **self._retry_policy):
+            with attempt:
+                if first_try:
+                    # on first try, we use the initial iterator.
+                    first_try = False
+                else:
+                    # Error retry reattach: After an error, attempt to
+                    self._iterator = self._stub.ReattachExecute(
+                        self._create_reattach_execute_request()
+                    )
+                ret = self._current
+
+        if ret is None:
+            raise StopIteration()
+
+        assert ret is not None
+        self._last_response_id = ret.response_id
+        if ret.response_complete:
+            self._response_complete = True
+            self._release_execute(None)  # release all
+        else:
+            self._release_execute(self._last_response_id)
+        self._current = None
+        return ret
+
+    def _has_next(self) -> bool:
+        if self._response_complete:
+            # After response complete response
+            return False
+        else:
+            for attempt in Retrying(
+                can_retry=SparkConnectClient.retry_exception, **self._retry_policy
+            ):
+                with attempt:
+                    if self._current is None:
+                        try:
+                            self._current = next(self._iterator)
+                        except StopIteration:
+                            pass
+                    has_next = self._current is not None
+
+                    # Graceful reattach:
+                    # If iterator ended, but there was no ResponseComplete, it means that
+                    # there is more, and we need to reattach. While ResponseComplete didn't
+                    # arrive, we keep reattaching.
+                    if not has_next and not self._response_complete:
+                        while not has_next:
+                            self._iterator = self._stub.ReattachExecute(
+                                self._create_reattach_execute_request()
+                            )
+                            # shouldn't change
+                            assert not self._response_complete
+                            try:
+                                self._current = next(self._iterator)
+                            except StopIteration:
+                                pass
+                            has_next = self._current is not None
+                    return has_next
+            return False
+
+    def _release_execute(self, until_response_id: Optional[str]) -> None:
+        request = self._create_release_execute_request(until_response_id)
+        for attempt in Retrying(can_retry=SparkConnectClient.retry_exception, **self._retry_policy):
+            with attempt:
+                self._stub.ReleaseExecute(request)
```

Comment:
is it possible to make the ReleaseExecute async - the idea is that client just "fires-and-forget" them as help to server to not do too much buffering, but doesn't get blocked on them.